gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java

```diff
@@ -132,16 +144,65 @@ public static void waitJobCompletion(
       WorkflowContext workflowContext = TaskDriver.getWorkflowContext(helixManager, workFlowName);
       if (workflowContext != null) {
-        org.apache.helix.task.TaskState helixJobState = workflowContext.getJobState(TaskUtil.getNamespacedJobName(workFlowName, jobName));
-        if (helixJobState == org.apache.helix.task.TaskState.COMPLETED ||
-            helixJobState == org.apache.helix.task.TaskState.FAILED ||
-            helixJobState == org.apache.helix.task.TaskState.STOPPED) {
+        switch (workflowContext.getJobState(TaskUtil.getNamespacedJobName(workFlowName, jobName))) {
+          case STOPPED:
+            // user requested cancellation, which is executed by executeCancellation()
+            log.info("Job {} is cancelled, it will be deleted now.", jobName);
+            new TaskDriver(helixManager).deleteAndWaitForCompletion(workFlowName, 10000L);
+            return;
+          case FAILED:
+          case COMPLETED:
           return;
         }
+      } else {
+        // We have waited for WorkflowContext to get initialized,
+        // so it is found null here, it must have been deleted in job cancellation process.
+        log.info("WorkflowContext not found. Job is probably cancelled.");
+        return;
       }
+      log.info("Waiting for job {} to complete...", jobName);
       Thread.sleep(1000);
     }
 
     throw new TimeoutException("task driver wait time [" + timeoutInSeconds + " sec] is expired.");
   }
-}
+
+  public static void handleJobTimeout(String workFlowName, String jobName, HelixManager helixManager,
+      Object jobLauncher, JobListener jobListener) throws InterruptedException {
+    try {
+      if (jobLauncher instanceof GobblinHelixJobLauncher) {
+        ((GobblinHelixJobLauncher) jobLauncher).cancelJob(jobListener);
+      } else if (jobLauncher instanceof GobblinHelixDistributeJobExecutionLauncher) {
+        ((GobblinHelixDistributeJobExecutionLauncher) jobLauncher).cancel();
+      }
+    } catch (JobException e) {
+      throw new RuntimeException("Unable to cancel job " + jobName + ": ", e);
+    }
+    // TODO : fix this when HELIX-1180 is completed
+    // We should not be deleting a workflow explicitly.
+    // Workflow state should be set to a final state, which will remove it automatically because expiry time is set.
+    // After that, all delete calls can be replaced by something like HelixUtils.setStateToFinal();
+    HelixUtils.deleteStoppedHelixJob(helixManager, workFlowName, jobName);
+    log.info("Stopped and deleted the workflow {}", workFlowName);
+  }
+  /**
+   * Deletes the stopped Helix Workflow.
+   * Caller should stop the Workflow before calling this method.
+   * @param helixManager helix manager
+   * @param workFlowName workflow needed to be deleted
+   * @param jobName helix job name
+   * @throws InterruptedException
+   */
+  public static void deleteStoppedHelixJob(HelixManager helixManager, String workFlowName, String jobName)
```

Comment:
Do we want to force user to follow the stop_and_delete approach? If that is the case, this method should be private?